### PR TITLE
fixed #305 and #306

### LINKED ISSUE #305: Data-Table: Issue with page-size-selector for infinite feeds
When filteredCount is undefined the pageSizeSelector is always visible but has no values irrespective of the pageSizeOptions.

### PR DESCRIPTION
**Pull request by jonthans**
resolves #305 and #306

### CODE REVIEW

File: demo/index.html [hunk 1]
```diff
@@ -831,27 +831,35 @@
...
-                    id: 1,
-                    cells: { 'name': 'Jan', 'age': 41, 'profession': 'Artist' }
-                  },
-                  {
-                    id: 2,
-                    cells: { 'name': 'Dan', 'age': 41, 'profession': 'Builder' }
-                  }
-                ]
-              })
+              feed: {
```
Comment (jonthans) [inline demo/index.html]: updated the example to use a "lazy feed"

File: modules/data-table/main/index.coffee [hunk 2]
```diff
@@ -663,8 +663,6 @@ class DataTable extends hx.EventEmitter
           start = undefined
           end = undefined
 
-        selection.classed('hx-data-table-infinite', totalCount is undefined)
```
Comment (jonthans) [inline modules/data-table/main/index.coffee]: allow for the table to change whether it is still considered infinite or not by checking for that at a later point

File: modules/data-table/main/index.coffee [hunk 3]
```diff
@@ -746,22 +746,21 @@ class DataTable extends hx.EventEmitter
               @_.sortColPicker.value({value: currentSort.column + currentSort.direction})
 
           # populate the page size picker if there are options set
-          if filteredCount isnt undefined and filteredCount > 0
```
Comment (jonthans) [inline modules/data-table/main/index.coffee]: remove this if as the filteredCount seems unrelated to the block it is surrounding.
The problem with this is that it makes the pageSizeOptions always get rendered without any items

File: modules/data-table/main/index.coffee [hunk 4]
```diff
@@ -677,6 +675,8 @@ class DataTable extends hx.EventEmitter
         feed.rows range, ({rows, filteredCount}) =>
           if options.displayMode is 'paginate'
             multiPage = false
+            selection.classed('hx-data-table-infinite', filteredCount is undefined)
```
Comment (jonthans) [inline modules/data-table/main/index.coffee]: This is to prevent a seemingly orphaned "of <n>" shown without the pageSelector next to it.

**Comment by jonthans** [comment]
After this change the totalCount only gets used when the user selects rows. Maybe that means that the check for totalCount can be moved down into the @updateSelected method and the option on the feed can be made optional

File: demo/index.html (hunk 1 above)
Comment (jmsmyth) [inline demo/index.html]: Could we also keep the object feed example in as well? - so have both this one and the old version, just to make any manual testing a bit easier

File: demo/index.html [hunk 5]
```diff
@@ -831,27 +831,35 @@
...
+                      {
+                        id: 1,
+                        cells: { 'name': 'Jan', 'age': 41, 'profession': 'Artist' }
+                      },
+                      {
+                        id: 2,
+                        cells: { 'name': 'Dan', 'age': 41, 'profession': 'Builder' }
+                      }
+                    ].slice(range.start, range.end + 1),
+                    filteredCount: range.end >= 3 ? 3 : undefined
```
Comment (jmsmyth) [inline demo/index.html]: `range.end >= 3 ? 3 : undefined` should be `range.end >= 2 ? 3 : undefined`, otherwise the pagination breaks for the example

Comment (jonthans) [inline demo/index.html]: Well, the idea is that the table is able to see that the feed has ended and from thereon displays the final size. I have updated the example to hold on to the fact that the size is known as soon as `range.end >= 3` so that when you go back to only showing the first 2 rows you will see `1-2  of 3` in the paginator

Comment (jmsmyth) [inline demo/index.html]: Just to clarify, the problem here is that when the page size cleanly divides the filtered count you can click the next button even when no data exists for the next page:

data-table-pagination-example

The same happens if the page size is set to 3.

I think this is simply because range.end uses 0 based indexing, so `range.end >= 3` is actually checking if we've gone beyond the end of the 'stream' by 1, rather than checking if its at end. So for this example checking `range.end >=2` would resolve the problem. (If I've misunderstood how it works, please correct me :-) )

This would be the same in the iterator/stream situation, where you could check something like `iterator.hasNext()` before deciding to return a value for `filteredCount` to avoid the empty page problem.

Comment (jonthans) [inline demo/index.html]: gotcha :+1:

**Comment by jmsmyth** [comment]
This will be released in 1.7.0
